ampd/src/mvx/verifier.rs

```diff
@@ -0,0 +1,620 @@
+        if event_name.as_slice() != CONTRACT_CALL_EVENT.as_bytes() {
+            return Ok(false);
+        }
+
+        let sender = topics.get(1).ok_or(Error::PropertyEmpty)?;
+        let sender = STANDARD.decode(sender)?;
+        if sender.len() != 32 || &sender[0..32] != &self.source_address.to_bytes() {
+            return Ok(false);
+        }
+
+        let destination_chain = topics.get(2).ok_or(Error::PropertyEmpty)?;
+        let destination_chain = STANDARD.decode(destination_chain)?;
+        let destination_chain = String::from_utf8(destination_chain)?;
+        if destination_chain != self.destination_chain.to_string() {
+            return Ok(false);
+        }
+
+        let destination_address = topics.get(3).ok_or(Error::PropertyEmpty)?;
+        let destination_address = STANDARD.decode(destination_address)?;
+        let destination_address = String::from_utf8(destination_address)?;
+        if destination_address != self.destination_address {
+            return Ok(false);
+        }
+
+        let payload_hash = topics.get(4).ok_or(Error::PropertyEmpty)?;
+        let payload_hash = STANDARD.decode(payload_hash)?;
+        if payload_hash.len() != 32
+            || Hash::from_slice(payload_hash.as_slice()) != self.payload_hash
+        {
+            return Ok(false);
+        }
+
+        return Ok(true);
+    }
+}
+
+impl VerifierSetConfirmation {
+    fn eq_event(&self, event: &Events) -> Result<bool, Box<dyn std::error::Error>> {
+        if event.identifier != ROTATE_SIGNERS_IDENTIFIER {
+            return Ok(false);
+        }
+
+        let topics = event.topics.as_ref().ok_or(Error::PropertyEmpty)?;
+
+        let event_name = topics.get(0).ok_or(Error::PropertyEmpty)?;
+        let event_name = STANDARD.decode(event_name)?;
+        if event_name.as_slice() != SIGNERS_ROTATED_EVENT.as_bytes() {
+            return Ok(false);
+        }
+
+        let signers_hash = topics.get(2).ok_or(Error::PropertyEmpty)?;
+        let signers_hash = STANDARD.decode(signers_hash)?;
+
+        let weighted_signers = WeightedSigners::from(&self.verifier_set);
+
+        if signers_hash.len() != 32 || signers_hash.as_slice() != weighted_signers.hash().as_slice()
+        {
+            return Ok(false);
+        }
+
+        return Ok(true);
+    }
+}
+
+fn find_event<'a>(
+    transaction: &'a TransactionOnNetwork,
+    gateway_address: &Address,
+    log_index: usize,
+    identifier: &str,
+    needed_event_name: &[u8],
+) -> Option<&'a Events> {
+    // We only support log index being 0, so one supported event per transaction because of MultiversX limitations
+    if log_index != 0 {
+        return None;
+    }
+
+    // Because of current relayer limitation, if log_index is 0, we will search through the logs and get the first log
+    // which corresponds to the gateway address, hence only supporting one cross chain call per transaction
+    transaction
+        .logs
+        .as_ref()?
+        .events
+        .iter()

```

Comment:
why can't you just add an index to these events, and then properly use the event index of the message id?

Reply:
The index problem is not here (on amplifier side), but rather on MUltiversX Event Notifier service side which is used by the Relayer. When events come there, they don't have any log index (so no order from the transaction), hence we only know that a `callContract` event happened, we know the txHash but we don't know the actual log index of that event, and this is not going to be fixed in the near future as far as I know.

Reply:
Does multiversx support multiple contract call events in the same transaction though? If so, what would happen if a single transaction calls the gateway more than once and thus generates more than one cross chain message? It looks like only the first will be routed.

Reply:
Yes, only the first will be routed, it is a limitation that applications should keep in mind since there is no way to limit this on the MultiversX contract side.

Reply:
we need to find a way to either route them, or reject those later events at the gateway. The gateway should not accept msgs that cannot be routed

Reply:
also, is the order of events here deterministic? If so, can we just assign indexes in ampd? If not, I think that's a larger problem, as different ampd nodes might see a different order, and thus not agree on the first event in the transaction

Reply:
Yeah, this would limit what apps can do severely. In fact, for ITS we allow a user to deploy multiple tokens within the same tx via a multicall (i.e multiple contract calls). Other apps might need to perform a broadcast as well. I would like to know what the limitation is on having unique event indexes for the contract call event. If it's not possible to uniquely identify indices, then I recommend using a counter in the external gateway and including that in the event for uniqueness.

Reply:
@cjcobb23 @milapsheth TLDR: We will add support for event index in ampd.

The reason we didn't add this initially is because of limitations on the Relayer/Event Notifier side for MultiversX, since the order is not supported there.

However, the event order is supported on the endpoints we use in ampd. To answer your questions:

1. There is no way to limit the events on Mvx Gateway contract side, the Gateway supports multiple cross chain contract calls per transaction and there is no limitation here.
Furthermore, the Gateway contract code is finished and is very close to the Solidity implementation, hence we will not do any other changes there

2. The order of events is deterministic, so we can indeed assign indexes in ampd. This was one of the initial implementations we had, I will re-add the necessary code and support event indexes for MultiversX

3. I can see how having multiple cross chain calls from MultiversX is useful, and we will support these. 

The only issue remains on Relayer side, there is a workaround that is not that efficient but we will figure out another solution.

Our priority is to get the ampd code merged and then we can handle the Relayer. I will do the necessary changes ASAP.